doc/api/n-api.md

```diff
@@ -0,0 +1,3024 @@
+
+Function pointer used with functions that support asynchronous
+operations. Callback functions must statisfy the following signature:
+
+```C
+typedef void (*napi_async_execute_callback)(napi_env env, void* data);
+```
+
+#### napi_async_complete_callback
+
+Function pointer used with functions that support asynchronous
+operations. Callback functions must statisfy the following signature:
+
+```C
+typedef void (*napi_async_complete_callback)(napi_env env,
+                                             napi_status status,
+                                             void* data);
+```
+
+## Error Handling
+N-API uses both return values and Javascript exceptions for error handling.
+The following sections explain the approach for each case.
+
+### Return values
+
+All of the N-API functions share the same error handling pattern. The
+return type of all API functions is `napi_status`.
+
+The return value will be `napi_ok` if the request was successful and
+no uncaught JavaScript exception was thrown. If an error occurred AND
+an exception was thrown, the `napi_status` value for the error
+will be returned. If an exception was thrown, and no error occurred,
+`napi_pending_exception` will be returned.
+
+In cases where a return value other than `napi_ok` or
+`napi_pending_exception` is returned, [`napi_is_exception_pending`][]
+must be called to check if an exception is pending.
+See the section on exceptions for more details.
+
+The full set of possible napi_status values is defined
+in `napi_api_types.h`.
+
+The `napi_status` return value provides a VM-independent representation of
+the error which occurred. In some cases it is useful to be able to get
+more detailed information, including a string representing the error as well as
+VM (engine)-specific information. 
+
+**Note:** Do not rely on the content or format of any of the extended
+information as it is not subject to SemVer and may change at any time.
+It is intended only for logging purposes.
+
+In order to retrieve this information, the following method is provided:
+
+```C
+NODE_EXTERN const napi_extended_error_info* napi_get_last_error_info();
+```
+The format of the `napi_extended_error_info` structure is as follows:
+
+```C
+struct napi_extended_error_info {
+  const char* error_message;
+  void* engine_reserved;
+  uint32_t engine_error_code;
+  napi_status error_code;
+};
+```
+
+[`napi_get_last_error_info`][] returns the information for the last
+N-API call that was made. The `error_code` will match the
+`napi_status` returned by that call. The `error_message` will
+be a textual representation of the error that occurred.
+
+
+#### napi_get_last_error_info
+<!-- YAML
+added: v8.0.0
+-->
+```C
+NAPI_EXTERN napi_status
+napi_get_last_error_info(napi_env env,
+                         const napi_extended_error_info** result);
+```
+- `[in]  env`: The environment that the API is invoked under
+- `[out]  result`: The `napi_extended_error_info` structure with more
```

Comment:
Can we use 1 space between `[in/out]` and `name`? Here and throughout. They don't render with more than 1 space it seems anyways?

(Unless `make doc` does, but I doubt it?)